test: add a document describing Cockpit's test Architecture (#19177)
cockpit-project/cockpit: test/ARCHITECTURE.md

    def testBasic(self):
        self.login_and_go("/system")
        self.machine.execute("echo 'test' > /etc/motd")
        self.browser.wait_text("#motd", 'test')

if __name__ == '__main__':
    testlib.test_main()
```

In Cockpit there are two types of tests, "destructive" and "nondestructive".
tests. Destructive tests do something to the virtual machine which makes it
unable to run another test afterwards or requires another virtual machine for
testing. The test above is a "destructive" test which is the default, a non
"destructive" test makes sure any destructive action is restored after the test
has run as can be seen below. Nondestructive tests were introduced to be able
to run them on a single running machine, such as the Testing Farm or a custom VM.
They also speed up testing, as booting a machine for every tests incurs a
significant penalty of ~ 10-30 seconds per test.

```python
@testlib.nondestructive
class TestSystem(testlib.MachineCase):
    def testBasic(self):
        self.login_and_go("/system")
        self.write_file("/etc/motd", "test")
        self.browser.wait_text("#motd", 'test')

if __name__ == '__main__':
    testlib.test_main()
```

> The test above would be invoked via `./test/verify/check-apps TestApps.testBasic`

Reviewer: The class name in the example is `TestSystem` so `./test/verify/check-apps TestSystem.testBasic`?

and would execute as can be seen in the diagram below:
```mermaid
sequenceDiagram
    participant test
    participant machine
    participant browser

    test->>test: test_main()
    test->>test: setUp()
    test->>machine: start()
    test->>machine: wait_boot()
    test->>browser: __init__()
    test->>test: setup non-destructive setup
    test->>test: run test
    test->>browser: start()
    test->>test: tearDown()
    test->>browser: kill()
    test->>machine: kill()
```

A test starts by calling `test_main`, this provides common command line
arguments for debugging and to optionally run a test on a different
machine/browser. These arguments are available in the `MachineCase` class as
`opts`. `test_main` also takes care of instantiating a `TapRunner` which runs
all the specified tests sequentially.

Once a test is started it runs `MachineCase.setUp` which has the responsibility

> to start a virtual machines(s) depending on if it is a "non-destructive" or

Reviewer: "virtual machines(s)" - extra "s" (?)

    test failures"]
    testfailed["Test failed"]
    failure_policy{"Known issue?"}
    
    finished --> skipped
    skipped --> |No| succeeded
    skipped --> |Yes| skiptest[Show as skipped test]
    succeeded --> |No| todonosuccess
    succeeded --> |yes| affected
   
    affected --> |Yes| affecteddone
    affected --> |No| todosuccess
    
    todosuccess --> |Yes| todosucceeded
    todosuccess --> |No| done[Test succeeded]
    
    todonosuccess --> |Yes| todofail
    todonosuccess --> |No| failure_policy
    
    failure_policy --> |Yes| known_issue[Show as known issue]
    failure_policy --> |No| pixel_journal_failure
    
    pixel_journal_failure --> |Yes| testfailed
    pixel_journal_failure --> |No| retry
```

* `Skipped` -  tests can be skipped because the test can not run on the given `TEST_OS`.
* `Affected` - tests to be retried to make sure any changes affecting them do not lead to flaky tests.
* `Todo` - tests which are incomplete and expected to fail.
* `Known issue` - Naughties are expected test failures due to known issues in software we test,
                  we still run the test but if the test error output matches a known naughty it

>                   is skipped. The [bots](github.com/cockpit-project/bots) repository keeps track

Reviewer: `https://` before "github.com/cockpit-project/bots", the link doesn't work otherwise.

# Architecture

This document describes the architecture of Cockpit's browser integration
tests. The tests should automate how a normal user interacts with Cockpit.
This requires a test machine which can add multiple disks or interfaces,
reboot, interact with multiple machines on the same network and run potentially

> destructive test scenario's (e.g. installing/updating packages, formatting

Reviewer 2: "scenarios"

disks).

For these reasons, Cockpit tests run inside a virtual machine (VM). The virtual
machine uses Cockpit specific virtual machine images maintained and created in
the [bots](https://github.com/cockpit-project/bots) repository. The images are
usually based on a distribution's cloud image customized with:

> * A well-known password for the admin/root user
> * Test SSH keys for access

Reviewer 2: ```suggestion
 * A well-known password and SSH key for the admin and root users
```

> * Test packages required to test Cockpit

Reviewer 2: ```suggestion
 * Packages required to test cockpit
```

* A build chroot with Cockpit's build dependencies to build the to be tested
  Cockpit source inside the virtual machine offline. This allows a developer on
  Fedora to easily prepare a Debian test image without having to install Debian
  build tools.
* Disabling system services which interfere with testing

To automate user actions on the web page, Cockpit is tested in a browser
controlled using the [Chrome DevTools Protocol](https://chromedevtools.github.io/devtools-protocol/)
(CDP) which is supported by Firefox and Chromium based browsers.

The Python test framework in [test/common](./common) is responsible for setting up the
test environment, running tests and reporting of the test output.

Diagram of the interaction of Browser/Machine/CDP/Test Framework.
```mermaid
graph TD;
    id[Test Framework] <-->|CDP| Browser;
    Browser <-->|HTTPS/WebSocket| Cockpit;

>     id1[Virtual Machine] <-->|HTTP/DBus/OS API's| Cockpit;

Reviewer 2: APIs

    id[Test Framework] <-->|SSH| id1[Virtual Machine];
```

## Integration Tests

Cockpit's tests can be run via three different entry points:

> * `test/verify/check-$page` - run a single or multiple unit test(s)

Reviewer 2: These are not unit tests, but integration. Just saying "test(s)" is fine.

* `test/common/run-tests` - run tests through our test scheduler (retries, tracks naughties, parallelism)
* `test/run` - CI entry point

We will start with how a single integration test is run and then explore the test
scheduler and CI setup.

### Test runtime

The base of a Cockpit integration test looks as following:

```python

> class TestApps(testlib.MachineCase):

Reviewer 2: Little weird to call this TestApps when it is exercising the /system page

    def testBasic(self):
        self.machine.execute("rm /etc/machine-id")
        self.login_and_go("/system")

>         self.browser.wait_text("#system_machine_id", "")

Reviewer 2: Note that this is a bad example: Immediately after the page loads, the chance is high that the value starts out empty anyway. A better pattern is to check for some non-empty string (e.g. actually read the current ID), then rm it, and then wait for empty.

if __name__ == '__main__':
    testlib.test_main()
```

> In Cockpit there are two types of tests, 'destructive" and "nondestructive".

Reviewer 2: Inconsistent quoting.

tests. Destructive tests do something to the virtual machine which makes it
unable to run another test afterwards or requires another virtual machine for
testing. The test above is a `destructive` test which is the default, a non

> `destructive` test makes sure any destructive action is restored after the test

Reviewer 2: non `destructive` is weird -- first the conjunction with "non", and also because this isn't actually a valid decorator. non-destructive or `@nondestructive`

> has run as can be seen below. Nondestructive tests where introduced to be able

Reviewer 2: "were"

to run them on a single running machine, such as the Testing Farm or a custom VM.

> They also speed up testing, as rebooting and shutting down a machine for every tests incurs a

Reviewer 2: We don't actually shut down machines at the end of the test, they just get killed, which takes a negligible time.

significant penalty of ~ 10-30 seconds per test.

```python
@testlib.nondestructive
class TestApps(testlib.MachineCase):
    def testBasic(self):
        self.restore_file("/usr/share/cockpit/apps")
        self.machine.execute("rm /etc/machine-id")
        self.login_and_go("/system")

>         self.browser.wait_text("#system_machine_id", "")

Reviewer 2: This is very incoherent. restore_file() does not work with a directory, it rms a different file than it (tries to) restore, it checks /system but is called TestApps, etc.

if __name__ == '__main__':
    testlib.test_main()
```

The test above would be invoked via `./test/verify/check-apps TestApps.testBasic`
and would execute as can be seen in the diagram below:
```mermaid
sequenceDiagram
    participant test
    participant machine
    participant browser

    test->>test: test_main()
    test->>test: setUp()
    test->>machine: start()
    test->>machine: wait_boot()
    test->>browser: __init__()
    test->>test: setup non-destructive setup
    test->>test: run test
    test->>browser: start()
    test->>test: tearDown()
    test->>browser: kill()
    test->>machine: kill()
```

A test starts by calling `test_main`, this provides common command line
arguments for debugging and to optionally run a test on a different
machine/browser. These arguments are available in the `MachineCase` class as
`opts`. `test_main` also takes care of instantiating a `TapRunner` which runs
all the specified tests sequentially.

Once a test is started it runs `MachineCase.setUp` which has the responsibility

> to start a virtual machines(s) depending on if it is a `nondestructive` or

Reviewer 2: As discussed above, better turn most of these destructive/nondestructive to normal words. You should of course explain/mention `@nondestructive` somewhere.

`destructive` test. If we run a `nondestructive` test a global machine is
created, and re-used for other `nondestructive` tests which might run. For

> `destructive` tests a machine is created on-demand, possible multiple machines

Reviewer 2: ```suggestion
destructive tests one or multiple machines are created depending on the test class'es `provision` value.
```

depending on the test class `provision` variable.

 For `nondestructive` tests cleanup handlers are installed to restore files in
 `/etc`, cleans up home directories etc.

 Lastly a `Browser` class is instantiated, this does not start the Browser
 directly but builds the required command for the `TEST_BROWSER` to start
 either Chromium or Firefox. When a test calls any method on the browser object
 the browser will be started, so tests which require no browser don't start a
 browser needlessly.

The `CDP` class is responsible for spawning the browser, then spawning a CDP

> driver this uses the `chrome-remote-interface` npm module to send commands to

Reviewer 2: Period after "driver", runaway sentence.

the spawned drivers via standard in (stdin).

On `tearDown` the test status is inspected, if it failed test logs are
collected and if the user has passed `--sit` the test pauses execution until
the user presses enter so that the machine/browser state can be inspected. The
test browser is killed after the `tearDown` function completed.

Virtual machines are killed by the `TapRunner` once all tests have finished or
in `setUp` if it's a `destructive` test as `nondestructive` tests re-use the
existing global machine.

### Test runner

Cockpit uses a custom test runner to run the tests, spread the load over jobs
and special handling of test failures. The test runner is implemented in
Python in [test/common/run-tests](./common/run-tests) and expects a list of tests to be provided.

The provided tests are collected and split up in `destructive` and

> `nondestrictive` tests and initialized as `Test` object. If there are any

Reviewer 2: destr*i*ctive typo

changes compared to the `main` branch, the test runner checks if any of the

> tests changed if so they are added to the affected test list unless more then

Reviewer 2: semicolon after "changed", runaway sentence.

three tests are changed. If `pkg/apps` is changed, `test/verify/check-apps`
will also be added to the affected test list 

After having collected the destructive, nondestructive and affected tests a
scheduling loop is started, if a machine was provided it is used for the

> nondestructive tests, destructive tests will always spawn a new machine. If no

Reviewer 2: Please split into three sentences.

machine is provided a pool of global machines is created based on the provided
`--jobs` and nondestructive tests. The test runner will first try to assign all
nondestructive tests on the available global machines and start the tests. 

> A test is started by the `Test` class by calling the `start()` method executed

Reviewer 2: I feel this is too much translating of Python to English, i. e. too much detail for an architecture overview. I think it'd be more interesting to see the general idea about how tests are scheduled and parallelized. Your call, though.

the provided `command` (e.g. `./test/verify/check-apps --machine 127.0.0.1:2201
--browser 127.0.0.1:9091`) with a `timeout` to cancel hanging tests after a

> timeout automatically and creates a temporarily file to store the results of

Reviewer 2: perhaps results → output?

the test `command` in. Finally the test is added to the `running_tests` list.

The test runner inspects all `running_tests` in a tight loop and polls if the
test is still running. If the process stopped or exited the test output and
`returncode` is saved. Depending on the `returncode` the test runner makes a

> decision on what to do next as described in the diagram below.

Reviewer 3: Hi Jelle,

I believe this should be rephrased as:

```markdown
decision on what to do next as described in the [flow chart](https://xkcd.com/518/) below:
```

;)

```mermaid
graph TD;
    finished[Test finished]
    succeeded{Test succeeded?}
    affected{"Test affected?"}
    affecteddone["Retry three times"]

>     skipped{Test skipped?}   

Reviewer 2: This looks confusing. Normally "skipped" means to not run a test at all. It's also not naughty matching, as the test succeeded, and there's a separate "known issues" branch. According to the explanation below, the "skipped" branch should come first.

# Architecture

This document describes the architecture of Cockpit's browser integration
tests. The tests should automate how a normal user interacts with Cockpit.
This requires a test machine which can add multiple disks or interfaces,
reboot, interact with multiple machines on the same network and run potentially
destructive test scenarios (e.g. installing/updating packages, formatting
disks).

For these reasons, Cockpit tests run inside a virtual machine (VM). The virtual
machine uses Cockpit specific virtual machine images maintained and created in
the [bots](https://github.com/cockpit-project/bots) repository. The images are
usually based on a distribution's cloud image customized with:

* A well-known password and SSH key for the admin and root users
* Packages required to test Cockpit
* A build chroot with Cockpit's build dependencies to build the to be tested

>   Cockpit source inside the virtual machine offline. This allows a developer on

Reviewer: The sentence "A build chroot with Cockpit's build dependencies to build the to be tested
  Cockpit source inside the virtual machine offline." sounds confusing.

Reviewer: Ok, after reading it a dozen times, it kind of makes sense now :)

  Fedora to easily prepare a Debian test image without having to install Debian
  build tools.
* Disabling system services which interfere with testing

To automate user actions on the web page, Cockpit is tested in a browser
controlled using the [Chrome DevTools Protocol](https://chromedevtools.github.io/devtools-protocol/)
(CDP) which is supported by Firefox and Chromium based browsers.

The Python test framework in [test/common](./common) is responsible for setting up the
test environment, running tests and reporting of the test output.

Diagram of the interaction of Browser/Machine/CDP/Test Framework.
```mermaid
graph TD;
    id[Test Framework] <-->|CDP| Browser;
    Browser <-->|HTTPS/WebSocket| Cockpit;
    id1[Virtual Machine] <-->|HTTP/DBus/OS APIs| Cockpit;
    id[Test Framework] <-->|SSH| id1[Virtual Machine];
```

## Integration Tests

Cockpit's tests can be run via three different entry points:

* `test/verify/check-$page` - run a single or multiple test(s)
* `test/common/run-tests` - run tests through our test scheduler (retries, tracks naughties, parallelism)

> * `test/run` - CI entry point

Reviewer 2: ```suggestion
* `test/run` or `.cockpit-ci/run` - CI entry point
```

We will start with how a single integration test is run and then explore the test
scheduler and CI setup.

### Test runtime

The base of a Cockpit integration test looks as following:

```python
class TestSystem(testlib.MachineCase):
    def testBasic(self):
        self.login_and_go("/system")
        machine_id = self.machine.execute("cat /etc/machine-id")
        self.browser.wait_text("#system_machine_id", machine_id)
        self.machine.execute("rm /etc/machine-id")
        self.browser.reload()

>         self.browser.wait_not_present("#system_machine_id")

Reviewer 2: We should avoid reload()s as much as possible. If we don't react to machine-id changes (which is fine, it really shouldn't change), then perhaps do this with motd?

After a reload, the test "loses" the current frame, so will be in the shell. So the wait_not_present() is doubly wrong: wrong frame, and always true immediately after reload. motd without reload avoids both.

Same for the nond example below.

The provided tests are collected and split up in "destructive" and
"non-destructive" tests and initialized as `Test` object. If there are any
changes compared to the `main` branch, the test runner checks if any of the
tests changed: if so they are added to the affected test list unless more then
three tests are changed. If `pkg/apps` is changed, `test/verify/check-apps`
will also be added to the affected test list 

After having collected the destructive, nondestructive and affected tests a
scheduling loop is started, if a machine was provided it is used for the
nondestructive tests, destructive tests will always spawn a new machine. If no
machine is provided a pool of global machines is created based on the provided
`--jobs` and nondestructive tests. The test runner will first try to assign all
nondestructive tests on the available global machines and start the tests. 

A test is started by the `Test` class by calling the `start()` method executed
the provided `command` (e.g. `./test/verify/check-apps --machine 127.0.0.1:2201
--browser 127.0.0.1:9091`) with a `timeout` to cancel hanging tests after a

> timeout automatically and creates a temporarily file to store the results of

Reviewer 2: ```suggestion
timeout automatically and creates a temporarily file to store the output of
```

* `Todo` - tests which are incomplete and expected to fail.
* `Known issue` - Naughties are expected test failures due to known issues in software we test,
                  we still run the test but if the test error output matches a known naughty it
                  is skipped. The [bots](github.com/cockpit-project/bots) repository keeps track
                  of all our known naughties per distro. (The bots repository has automation
                  setup to see if a naughty is still affected and if not open a pull request to
                   drop it).
* `Failed` - tests can fail due to our test shared infrastructure, instead of
             letting the whole test run fail, we re-try them unless `--no-retry-fail` is
             passed.

### Continuous Integration (CI)

In CI we have two entry points, one for our tests which runs on our own
managed infrastructure by [cockpituous](https://github.com/cockpit-project/cockpituous/)
and one for tests which run on the [testing farm (TF)](https://docs.testing-farm.io/).

For our own managed infrastructure the entry point of the Cockpit tests is

> `test/run`. This bash script expects a `TEST_OS` environment variable to be set

Reviewer 2: ```suggestion
`test/run`, or alternatively, `.cockpit-ci/run`. This bash script expects a `TEST_OS` environment variable to be set
```

# Architecture

This document describes the architecture of Cockpit's browser integration
tests. The tests should replicate how a normal user interacts with Cockpit this

> requires a test machine which can easily add have multiple disks or interfaces,

Reviewer 2: ```suggestion
requires a test machine which can add multiple disks or interfaces,
```

reboot, interact with multiple machines on the same network and run potentially
destructive test scenario's (e.g. installing/updating packages, formatting
disks).

For these reasons, Cockpit tests run inside a virtual machine (VM). The virtual
machine uses Cockpit specific virtual machine images maintained and created in
the [bots](https://github.com/cockpit-project/bots) repository. The images are
usually based on a distribution's cloud image customized with:

* A well-known password for the admin/root user
* Test SSH keys for access
* Test packages required to test Cockpit
* A build chroot with Cockpit's build dependencies to build the to be tested
  Cockpit source inside the virtual machine offline. This allows a developer on
  Fedora to easily prepare a Debian test image without having to install Debian
  build tools.
* Disabling system services which interfere with testing

To replicate a user, Cockpit is tested in a browser controlled using the
[Chrome DevTools Protocol](https://chromedevtools.github.io/devtools-protocol/)
(CDP) which is supported by Firefox and Chromium based browsers.

> The Python test framework in `test/common` is responsible for setting up the

Reviewer 2: please turn this into a  link

test environment, running tests and reporting of the test output.

Diagram of the interaction of Browser/Machine/CDP/Test Framework.
```mermaid
graph TD;
    id[Test Framework] <-->|CDP| Browser;

>     id[Test Framework] <-->|SSH| id1[Virtual Machine];

Reviewer 2: perhaps a "cockpit" between browser and VM?

Author: That is a good suggestion. Not sure how we define the interaction between Cockpit and the VM though.

Author: I don't think what I wrote is ideal, need some more time to think of something better.

```

## Integration Tests

Cockpit's tests can be run via three different entry points:

* `test/verify/check-$page` - run a single or multiple unit test(s)

> * `test/common/run-tests` - run tests through our test scheduler (retries, tracks naughties)

Reviewer 2: ```suggestion
* `test/common/run-tests` - run tests through our test scheduler (retries, tracks naughties, parallelism)
```

> * `test/run` - run tests in Continuous Integration (CI)

Reviewer 2: ```suggestion
* `test/run` - CI entry point
```

> * `test/run` - run tests in Continuous Integration (CI)

Reviewer 2: This is also useful to turn into a link

We will start with how a single integration test is run and then explore the test
scheduler and CI setup.

### Test runtime

The base of a Cockpit integration test looks as following:

```python
class TestApps(testlib.MachineCase):
    def testBasic(self):

>         self.machine.execute("rm /usr/share/cockpit/apps")

Reviewer 2: This is a directory, so `rm` will fail. Perhaps make this a little more abstract/synthetic, with something like running `echo some setup here`? Further down I see you want to point out the difference between destructive and non-destructive. So perhaps keep and fix this example.

        self.browser.login_and_go("/apps")

>         self.browser.wait_not_visible("#apps")

Reviewer 2: The login will fail because /apps doesn't exist. Also, a wait_not_visible() is a no-op or race condition. I. e. this example is very confusing and broken. This needs to stay on the shell, check that some other menu item is visible, and *then* check that Apps doesn't exist.

if __name__ == '__main__':
    testlib.test_main()
```

> In Cockpit there are two types of tests, `destructive` and `nondestructive`

Reviewer 2: ```suggestion
In Cockpit there are two types of tests, 'destructive" and "nondestructive".
```

Same below. These are not CLI programs or literal Python identifiers or so, just word definitions.

> tests. Destructive tests do something to test virtual machine which makes it

Reviewer 2: ```suggestion
tests. Destructive tests do something to the virtual machine which makes it
```

unable to run another test afterwards or requires another virtual machine for
testing. The test above is a `destructive` test which is the default, a non
`destructive` test makes sure any destructive action is restored after the test

> has run as can be seen below. `nondestructive` tests where introduced to speed
> up testing, as rebooting and shutting down a machine for every tests incurs a
> significant penalty of ~ 10-30 seconds per test.

Reviewer 2: ```suggestion
has run as can be seen below. Nondestructive tests where introduced to be able
to run them on a single running machine, such as the Testing Farm or a custom VM.
They also speed up testing, as rebooting and shutting down a machine for every tests incurs a
significant penalty of ~ 10-30 seconds per test.
```

```python
@testlib.nondestructive
class TestApps(testlib.MachineCase):
    def testBasic(self):
        self.restore_dir("/usr/share/cockpit/apps")
        self.machine.execute("rm /usr/share/cockpit/apps")
        self.browser.login_and_go("/apps")
        self.browser.wait_not_visible("#apps")

if __name__ == '__main__':
    testlib.test_main()
```

The test above would be invoked via `./test/verify/check-apps TestApps.testBasic`
and would execute as can be seen in the diagram below:
```mermaid

> sequenceDiagram

Reviewer 2: Please checkout the rendered diagram, this comes out all wrong and garbled :cry:

    participant test
    participant machine
    participant browser

    test->>test: test_main()
    test->>test: setUp()
    test->>machine: start()
    test->>machine: wait_boot()
    test->>browser: __init__()
    test->>test: setup non-destructive setup
    test->>test: run test
    test->>browser: start()
    test->>test: tearDown()
    test->>browser: kill()
    test->>machine: kill()
```

A test starts by calling `test_main`, this provides common command line
arguments for debugging and to optionally run a test on a different
machine/browser. These arguments are available in the `MachineCase` class as
`opts`. `test_main` also takes care of instantiating a `TapRunner` which runs
all the specified tests sequentially.

Once a test is started it runs `MachineCase.setUp` which has the responsibility

> to start a virtual machines(s) depending on if it is a `nondestrcutive` or

Reviewer 2: nondiscursive! sondermotive! nontestturbine! argh this is hard :laughing:

 Lastly a `Browser` class is instantiated, this does not start the Browser
 directly but builds the required command for the `TEST_BROWSER` to start
 either Chromium or Firefox. When a test calls any method on the browser object
 the browser will be started, so tests which require no browser don't start a
 browser needlessly.

The `CDP` class is responsible for spawning the browser, then spawning a CDP
driver this uses the `chrome-remote-interface` npm module to send commands to
the spawned drivers via standard in (stdin).

On `tearDown` the test status is inspected, if it failed test logs are
collected and if the user has passed `--sit` the test pauses execution until
the user presses enter so that the machine/browser state can be inspected. The
test browser is killed after the `tearDown` function completed.

Virtual machines are killed by the `TapRunner` once all tests have finished or
in `setUp` if it's a `destructive` test as `nondestructive` tests re-use the
existing global machine.

### Test runner

Cockpit uses a custom test runner to run the tests, spread the load over jobs
and special handling of test failures. The test runner is implemented in

> Python in `test/common/run-tests` and expects a list of tests to be provided.

Reviewer 2: link?

> The provided tests are collected and split up in `serial` and `parallel` tests

Author: After https://github.com/cockpit-project/cockpit/pull/19175 lands, this should be revised.

Reviewer 2: It did land

* `Todo` - tests which are incomplete and expected to fail.
* `Known issue` - Naughties are expected test failures due to known issues in software we test,
                  we still run the test but if the test error output matches a known naughty it
                  is skipped. The [bots](github.com/cockpit-project/bots) repository keeps track
                  of all our known naughties per distro. (The bots repository has automation
                  setup to see if a naughty is still affected and if not open a pull request to
                   drop it).
* `Failed` - tests can fail due to our test shared infrastructure, instead of
             letting the whole test run fail, we re-try them unless `--no-retry-fail` is
             passed.

### Continuous Integration (CI)

In CI we have two entry points, one for our tests which runs on our own
managed infrastructure by [cockpituous](https://github.com/cockpit-project/cockpituous/)
and one for tests which run on the [testing farm (TF)](https://docs.testing-farm.io/).

For our own managed infrastructure the entry point of the Cockpit tests is

> `test/run` this bash script expects a `TEST_OS` environment variable to be set

Reviewer 2: ```suggestion
`test/run`. This bash script expects a `TEST_OS` environment variable to be set
```

> to determine what distribution to run the tests under and a `TEST_SCENARIO`

Reviewer 2: ```suggestion
to determine what distribution to run the tests under, and a `TEST_SCENARIO`
```

environment variable to determine the type of test. Currently we support these

> different scenario's:

Reviewer 2: ```suggestion
different scenarios:
```
It's plural, not genitive or a contraction.

* `devel` - runs tests with coverage enabled and generates a html file with
  coverage information
* `pybridge` - runs tests with the Python bridge (soon to be deprecated after all
             images are moved over to the Python bridge)
* `firefox` - runs tests using the Firefox browser instead of Chrome
* `networking` - runs all networking related tests
* `storage` - runs all storage related tests
* `expensive` - runs all expensive tests (usually tests which reboot/generate a new initramfs)

> * `other` - runs all non-networking/storage/expensive tests.

Reviewer 2: I feel like this is going to become quickly out of sync with test/run and the bots testmap. Please either move that to test/run and link to it from this documentation, or add a reminder comment to test/run to also update this documentation on scenario changes. TBH I prefer the former -- I feel like this document already has too much detail which is prone to getting out of sync with the code.

> Cockpit's tests are split up in scenario's to heavily parallelize our testing and

Reviewer 2: ```suggestion
Cockpit's tests are split up in scenarios to heavily parallelize our testing and
```

Author: Fixed this everywhere.

allow for faster retrying.

> The `test/run` prepares an Virtual machine image for the given `TEST_OS` and then

Reviewer 2: ```suggestion
`test/run` prepares an Virtual machine image for the given `TEST_OS` and then
```

runs the tests by calling `test/common/run-tests` with the provided tests.

> For the Testing Farm (TF) scenario, Packit is responsible for setting up the
> test infrastructure and the entry point is `test/browser/browser.sh`. On TF we

Reviewer 2: ```suggestion
The [tmt](https://tmt.readthedocs.io) test is defined in [test/browser/main.fmf](./browser/main.fmf). These
run on the Testing Farm (TF), triggered by [Packit](https://packit.dev/) for upstream PRs. On TF we
```

get a single virtual machine without a hypervisor so tests run on the virtual

> machine directly this also implies that only `destructive` tests can be run.

Reviewer 2: ```suggestion
machine directly. This also implies that only nondestructive tests can be run.
```

The `test/browser/browser.sh` script sets up the virtual machine and calls
`test/browser/run-tests.sh` which selects a subset of all the `nondestructive`
tests to run using `test/common/run-tests`.

> `TF` scenario's are also split up into scenario's (basic, networking, optional)

Reviewer 2: ```suggestion
The tmt test is also split up into three plans (basic, networking, optional)
```
